### PR TITLE
[TextCNN] Add TextCNN example for sentiment analysis

### PR DESCRIPTION
**Pull request by linjieccc**
### PR types

Others

### PR changes

Others

### Description

Add TextCNN example for sentiment analysis

### CODE REVIEW

File: examples/sentiment_analysis/textcnn/README.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+### 数据准备
+
+#### 使用PaddleNLP内置数据集
+
+这里我们提供一份已标注的机器人聊天数据集，包括训练集（train.tsv），开发集（dev.tsv）和测试集（test.tsv）。
+
+```python
+from paddlenlp.datasets import load_dataset
+
+train_ds, dev_ds, test_ds = load_dataset("robotchat", splits=["train", "dev", "test"])
```
Comment (ZeyuChen) [inline examples/sentiment_analysis/textcnn/README.md]: 这份数据集建议外挂，而不是作为官方数据集。

File: examples/sentiment_analysis/textcnn/README.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+├── export_model.py # 动态图参数导出静态图参数脚本
+├── model.py # 模型组网脚本
+├── predict.py # 模型预测脚本
+├── README.md # 文档说明
+└── train.py # 对话情绪识别任务训练脚本
+```
+
+### 数据准备
+
+#### 使用PaddleNLP内置数据集
```
Comment (Steffy-zxf) [inline examples/sentiment_analysis/textcnn/README.md]: 数据集建议修改为加载自定义数据集的方式。

File: examples/sentiment_analysis/textcnn/model.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        self.embedder = nn.Embedding(
+            vocab_size, emb_dim, padding_idx=padding_idx)
+        self.encoder = CNNEncoder(
+            emb_dim=emb_dim,
+            num_filter=num_filter,
+            ngram_filter_sizes=ngram_filter_sizes)
+        self.fc = nn.Linear(self.encoder.get_output_dim(), fc_hidden_size)
+        self.output_layer = nn.Linear(fc_hidden_size, num_classes)
+
+    def forward(self, text, seq_len=None):
```
Comment (Steffy-zxf) [inline examples/sentiment_analysis/textcnn/model.py]: textcnn不需要seq_len

File: examples/sentiment_analysis/textcnn/README.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+Data: 下周星期一就要考试了我害怕我考试不会考考的很差     Label: negative
+Data: 你看看我附近有什么好吃的   Label: neutral
+Data: 我喜欢画画也喜欢唱歌       Label: positive
+```
+
+## Reference
+
+TextCNN参考论文：
+
+- https://arxiv.org/pdf/1408.5882.pdf
```
Comment (Steffy-zxf) [inline examples/sentiment_analysis/textcnn/README.md]: 建议修改为
[EMNLP2014-Convolutional Neural Networks for Sentence Classification](https://aclanthology.org/D14-1181.pdf)

File: examples/sentiment_analysis/textcnn/README.md [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+- https://arxiv.org/pdf/1510.03820.pdf
```
Comment (Steffy-zxf) [inline examples/sentiment_analysis/textcnn/README.md]: 这篇不是TextCNN原论文，可以先不写

File: examples/sentiment_analysis/textcnn/model.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Steffy-zxf) [inline examples/sentiment_analysis/textcnn/model.py]: 2021

File: examples/sentiment_analysis/textcnn/predict.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+parser = argparse.ArgumentParser(__doc__)
+parser.add_argument('--device', choices=['cpu', 'gpu', 'xpu'], default="gpu", help="Select which device to train model, defaults to gpu.")
+parser.add_argument("--batch_size", type=int, default=1, help="Total examples' number of a batch for training.")
+parser.add_argument("--vocab_path", type=str, default="./robot_chat_word_dict.txt", help="The path to vocabulary.")
+parser.add_argument("--params_path", type=str, default='./checkpoints/final.pdparams', help="The path of model parameter to be loaded.")
+args = parser.parse_args()
+# yapf: enable
+
+def preprocess_prediction_data(data, tokenizer, pad_token_id=0, max_gram_filter_size=3):
+    """
```
Comment (Steffy-zxf) [inline examples/sentiment_analysis/textcnn/predict.py]: max_gram_filter_size从哪里来的，没有解释清楚

并且增加相应的api注释

File: examples/sentiment_analysis/textcnn/train.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+    # Load vocab.
+    if not os.path.exists(args.vocab_path):
+        raise RuntimeError('The vocab_path  can not be found in the path %s' %
+                           args.vocab_path)
+
+    vocab = Vocab.load_vocabulary(
+        args.vocab_path, unk_token='[UNK]', pad_token='[PAD]')
+
+    train_ds, dev_ds = load_dataset("robotchat", splits=["train", "dev"])
```
Comment (Steffy-zxf) [inline examples/sentiment_analysis/textcnn/train.py]: 这个样例数据集 没有 测试集吗？

File: paddlenlp/datasets/robotchat.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Steffy-zxf) [inline paddlenlp/datasets/robotchat.py]: 改成加载自定义数据集的方式

**Review by Steffy-zxf — CHANGES_REQUESTED** [review]
数据集的加载方式还需要修改。

File: examples/sentiment_analysis/textcnn/deploy/python/predict.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+
+import numpy as np
+import paddle
+from paddle import inference
+from paddlenlp.data import JiebaTokenizer, Stack, Tuple, Pad, Vocab
+from scipy.special import softmax
```
Comment (ZeyuChen) [inline examples/sentiment_analysis/textcnn/deploy/python/predict.py]: 为什么要从scipy引入softmax？

File: examples/sentiment_analysis/textcnn/model.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        self.output_layer = nn.Linear(fc_hidden_size, num_classes)
+
+    def forward(self, text, seq_len=None):
+        # Shape: (batch_size, num_tokens, embedding_dim)
+        embedded_text = self.embedder(text)
+        # Shape: (batch_size, len(ngram_filter_sizes) * num_filter)
+        encoder_out = self.encoder(embedded_text)
+        encoder_out = paddle.tanh(encoder_out)
+        # Shape: (batch_size, fc_hidden_size)
+        fc_out = paddle.tanh(self.fc(encoder_out))
```
Comment (ZeyuChen) [inline examples/sentiment_analysis/textcnn/model.py:59]: 代码的层次有点乱，建议一个layer一行。
不要一会合并一行，一会分两行。

File: examples/sentiment_analysis/textcnn/train.py [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+            dataset=dataset, batch_size=batch_size, shuffle=shuffle)
+    dataloader = paddle.io.DataLoader(
+        dataset, batch_sampler=sampler, collate_fn=batchify_fn)
+    return dataloader
+
+
+def convert_example(example, tokenizer, is_test=False):
+
+    input_ids = tokenizer.encode(example["text"])
+    valid_length = np.array(len(input_ids), dtype='int64')
```
Comment (ZeyuChen) [inline examples/sentiment_analysis/textcnn/train.py]: 需要len吗？

File: examples/sentiment_analysis/textcnn/train.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+    else:
+        sampler = paddle.io.BatchSampler(
+            dataset=dataset, batch_size=batch_size, shuffle=shuffle)
+    dataloader = paddle.io.DataLoader(
+        dataset, batch_sampler=sampler, collate_fn=batchify_fn)
+    return dataloader
+
+
+def convert_example(example, tokenizer, is_test=False):
+
```
Comment (ZeyuChen) [inline examples/sentiment_analysis/textcnn/train.py]: 删除无意义的空行

File: examples/sentiment_analysis/textcnn/README.md (hunk 2 above)
Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/README.md]: 已修改

File: examples/sentiment_analysis/textcnn/model.py (hunk 3 above)
Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/model.py]: 已修改

File: examples/sentiment_analysis/textcnn/README.md (hunk 4 above)
Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/README.md]: 已修改

File: examples/sentiment_analysis/textcnn/README.md (hunk 5 above)
Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/README.md]: 已删除

File: examples/sentiment_analysis/textcnn/model.py (hunk 6 above)
Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/model.py]: 已修改

File: examples/sentiment_analysis/textcnn/predict.py (hunk 7 above)
Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/predict.py]: 已修改并增加注释

File: examples/sentiment_analysis/textcnn/train.py (hunk 8 above)
Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/train.py]: 已添加

File: paddlenlp/datasets/robotchat.py (hunk 9 above)
Comment (linjieccc) [reply paddlenlp/datasets/robotchat.py]: 已修改

File: examples/sentiment_analysis/textcnn/README.md (hunk 1 above)
Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/README.md]: 已修改

File: examples/sentiment_analysis/textcnn/deploy/python/predict.py (hunk 10 above)
Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/deploy/python/predict.py]: 已修改为paddle.nn.functional.softmax

File: examples/sentiment_analysis/textcnn/model.py (hunk 11 above)
Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/model.py:59]: 已修改

File: examples/sentiment_analysis/textcnn/train.py (hunk 12 above)
Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/train.py]: 已删除length

File: examples/sentiment_analysis/textcnn/train.py (hunk 13 above)
Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/train.py]: 已修改

File: examples/sentiment_analysis/textcnn/data.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    It process the prediction data as the format used as training.
+    Args:
+        data (obj:`List[str]`): The prediction data whose each element is a tokenized text.
+        tokenizer(obj: paddlenlp.data.JiebaTokenizer): It use jieba to cut the chinese string.
+        pad_token_id(obj:`int`, optional, defaults to 0): The pad token index.
+        max_ngram_filter_size (obj:`int`, optional, defaults to 3) Max n-gram size in TextCNN model.
+            Users should refer to the ngram_filter_sizes setting in TextCNN, if ngram_filter_sizes=(1, 2, 3)
+            then max_ngram_filter_size=3
+    Returns:
+        examples (obj:`List(Example)`): The processed data whose each element is a Example (pad) object.
```
Comment (Steffy-zxf) [inline examples/sentiment_analysis/textcnn/data.py]: 注释对吗？Example是什么？

File: examples/sentiment_analysis/textcnn/data.py [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+        sampler = paddle.io.BatchSampler(
+            dataset=dataset, batch_size=batch_size, shuffle=shuffle)
+    dataloader = paddle.io.DataLoader(
+        dataset, batch_sampler=sampler, collate_fn=batchify_fn)
+    return dataloader
+
+def preprocess_prediction_data(data, tokenizer, pad_token_id=0, max_ngram_filter_size=3):
+    """
+    It process the prediction data as the format used as training.
+    Args:
```
Comment (Steffy-zxf) [inline examples/sentiment_analysis/textcnn/data.py:58]: 55行之后增加一空行

File: examples/sentiment_analysis/textcnn/deploy/python/predict.py [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+    def predict(self,
+                data,
+                label_map,
+                batch_size=1,
+                pad_token_id=0):
+        """
+        Predicts the data labels.
+
+        Args:
+            model (obj:`paddle.nn.Layer`): A model to classify texts.
```
Comment (Steffy-zxf) [inline examples/sentiment_analysis/textcnn/deploy/python/predict.py]: 注释不对，建议通篇检查下

File: examples/sentiment_analysis/textcnn/data.py (hunk 14 above)
Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/data.py]: 已修改

File: examples/sentiment_analysis/textcnn/data.py (hunk 15 above)
Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/data.py:58]: 已修改

File: examples/sentiment_analysis/textcnn/deploy/python/predict.py (hunk 16 above)
Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/deploy/python/predict.py]: 已修改

File: examples/sentiment_analysis/textcnn/predict.py [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+        probs = F.softmax(logits, axis=1)
+        idx = paddle.argmax(probs, axis=1).numpy()
+        idx = idx.tolist()
+        labels = [label_map[i] for i in idx]
+        results.extend(labels)
+    return results
+
+
+if __name__ == "__main__":
+    paddle.set_device(args.device.lower())
```
Comment (ZeyuChen) [inline examples/sentiment_analysis/textcnn/predict.py]: 如果是args用了choice的话，这里不需要lower

Comment (linjieccc) [reply examples/sentiment_analysis/textcnn/predict.py]: 已修改

**Review by Steffy-zxf — APPROVED** [review]
LGTM
